sql/schemachanger: remove alter table add column limitations with not null constraints

fqazi opened the pull request:
Previously we had the following limitations with ALTER TABLE ADD COLUMN:

1. We disallowed ADD COLUMN ... VIRTUAL NOT NULL since our implementation did not do validation properly allowing this statement to succeed when it should fail.
2. We disallowed  ADD COLUMN .... UNIQUE NOT NULL because there was a risk of data loss in our index implementation. This was resolved by allowing multiple table mutations in a single statement.

This patch enables both of these statements after adding validation inside the end to end testing to ensure that DML injected scenarios work correctly. This patch also enhances some of the output generated from the DML injection tests to make things easier to diagnose.

Fixes: #117647

fqazi committed:
sql/schemachanger: enable ADD COLUMN VIRTUAL NOT NULL

Previously, we disabled ADD COLUMN VIRTUAL NOT NULL in the
declarative schema changer because we didn't have proper support
for validating constraints yet. Since, we added support for alter
table commands including constraints we can now expand what is supported
for ADD COLUMN. To address this, this patch will remove the fallback and
add DML injection to ensure this scenario now works correctly.

Release note: None
Fixes: #117647

fqazi committed:
sql/schemachanger: enable ADD COLUMN UNIQUE NOT NULL

Previously, ALTER TABLE ADD COLUMN UNIQUE NOT NULL was not enabled in
the declarative schemachanger because our add column implementation did
properly validate the NOT NULL constraint. This was fixed with later
work adding constraint support in the declarative schemachanger, so
can be safely enabled again. To address this, this patch is going to
enable this feature and add end to end tests.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118519)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: improve end to end testing logging

Previously, the error related logs from the schema changer
end to end tests did not format out the stage keys into human
readable strings. To address this, this patch modifies the
code to format these keys properly.
Epic: none
Release note: None

rafiss reviewed (commented):
nice, cool that it worked out of the box

can you also add a test in this framework: https://github.com/cockroachdb/cockroach/blob/16ad46ad757bbee9df8530740879eac640979a3d/pkg/sql/schemachanger/dml_injection_test.go#L144

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118519#-:-NqE2_IR97Hp2BaMMD1N:b-glrh88)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118519)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Xiang-Gu)



<!-- Sent from Reviewable.io -->

fqazi commented:
@rafiss TFTR!

bors r+
